### ISSUE TITLE
Very slow in WSL listing lot of files in /Windows/System32

### ISSUE THREAD

**Issue by valerii15298** [nushell/nushell#9057]
### Describe the bug

Nushell is very slow in directories with lot of files

### How to reproduce

Open `nu` in WSL, for example by running `bash.exe` on Windows and then `nu` in WSL.
Go to `/mnt/c/Windows/System32` (`cd /mnt/c/Windows/System32`)
Run `ls | wc -l`
It basically hangs indefinitely.

If you install nu via homebrew repro can be one line like this:
```powershell
bash.exe -c "/home/linuxbrew/.linuxbrew/bin/nu -c 'ls /mnt/c/Windows/System32 | wc -l'"
```

But note that it works for directories with a small amount of files, for example like this:
```powershell
bash.exe -c "/home/linuxbrew/.linuxbrew/bin/nu -c 'ls / | wc -l'"
```

### Expected behavior

It should display the number of lines from `ls`, in my case it is `5012`. Also bash, powerhsell or any other shell displays this number immediately when I run `ls | wc -l`

### Configuration

| key                | value                                                             |
| ------------------ | ----------------------------------------------------------------- |
| version            | 0.79.0                                                            |
| branch             |                                                                   |
| commit_hash        |                                                                   |
| build_os           | linux-x86_64                                                      |
| build_target       | x86_64-unknown-linux-gnu                                          |
| rust_version       | rustc 1.68.2 (9eb3afe9e 2023-03-27) (built from a source tarball) |
| cargo_version      | cargo 1.68.1 (6feb7c9cf 2023-03-26)                               |
| build_time         | 2023-04-25 20:06:17 +00:00                                        |
| build_rust_channel | release                                                           |
| features           | default, zip                                                      |
| installed_plugins  |                                                                   |

**Comment by fdncred** [comment]
I wonder if it has something to do with accessing networked files? I thought /mnt/c was treated as a networked folder.

**Comment by ChrisDenton** [comment]
Yes, in WSL2 the Windows filesystems are accessed via a 9P file server. Which is slow. But it perhaps shouldn't be *that* slow. Although I guess if nushell is querying 5k files one by one then that might explain it?

**Comment by valerii15298** [comment]
>Also bash, powerhsell or any other shell displays this number immediately when I run ls | wc -l

@ChrisDenton but if I use bash inside wsl it shows output immediately... So I guess it is a problem with *Nushell* specifically

**Comment by fdncred** [comment]
Could be. We'd welcome a PR to help with performance.

**Comment by ChrisDenton** [comment]
Yes, the second part of my comment was attempting to suggest a reason why nushell specifically might be uniquely slow in this case.

**Comment by fdncred** [comment]
In case you didn't see #9089, it should make `ls` from WSL to windows a bit faster. However, there's a chance it won't land because threading `ls`, the way I did it, makes it non-deterministic, so then I added sorting, which people won't necessarily like, and that also requires the output to be collected which means it won't stream well.

But, I'm trying. 😄

**Comment by valerii15298** [comment]
@fdncred thank you!!!
It sounds cool.

Can you explain how for example ls from bash works and why it isn't slow? BCS it also gets as @ChrisDenton said network files via 9P file server or maybe differently?

**Comment by skyvine** [comment]
Briefly discussed threading with @fdncred in some other issues, moving here because of topical appropriateness.

From #7579::

> I have a closed PR here somewhere with the one I wrote but the problem was that it was non-deterministic in it's sort, which could turn off users and after I made it sort, i resorted to collecting which broke the streaming rules.

[for reference: PR is #9089]

It sounds like the adherence to the streaming rules has to do with more than just fitting the output into the PiplineData abstraction, because the pull request did that by "streaming" the `vec` after the sort. So I would assume that it's about allowing different phases of the pipeline to execute in parallel. Is that correct?

On an unrelated note, I do not understand why it is important that `ls` stays sorted internally. When I read the [quick tour](https://www.nushell.sh/book/quick_tour.html), I had the impression that nu was sticking closely to the "do one thing well" principle. So I would expect the Nu perspective to be that if a user wants their list to be sorted, that is what the `sort-by` command is for. If they always want it to be sorted that way by default, then they can use an alias. This avoids complexity in the `ls` implementation.
